utils/common/src/main/java/org/apache/brooklyn/util/net/ReachableSocketFinder.java

```diff
@@ -69,86 +65,113 @@ public ReachableSocketFinder(Predicate<? super HostAndPort> socketTester, Listen
+     * Returns an iterable of the elements in sockets that are reachable. The order of elements
+     * in the iterable corresponds to the order of the elements in the input, not the order in which their
+     * reachability was determined. Iterators are unmodifiable and are evaluated lazily.
+     *
+     * @param sockets The host-and-ports to test
+     * @param timeout Max time to try to connect to each ip:port
+     * @return An iterable containing all sockets that are reachable according to {@link #socketTester}.
+     * @throws NullPointerException  If sockets or timeout is null
+     * @throws IllegalStateException  If the sockets to test is empty
      */
-    private Optional<HostAndPort> tryReachable(Collection<? extends HostAndPort> sockets, Duration timeout) {
-        final AtomicReference<HostAndPort> reachableSocket = new AtomicReference<HostAndPort>();
-        final CountDownLatch latch = new CountDownLatch(1);
-        List<ListenableFuture<?>> futures = Lists.newArrayList();
+    public Iterable<HostAndPort> findOpenSocketsOnNode(final Iterable<? extends HostAndPort> sockets, Duration timeout) {
+        checkNotNull(sockets, "sockets");
+        checkState(!Iterables.isEmpty(sockets), "No hostAndPort sockets supplied");
+        checkNotNull(timeout, "timeout");
+        return Optional.presentInstances(tryReachable(sockets, timeout));
+    }
+
+    /**
+     * @return A lazily computed Iterable containing present values for the elements of sockets that are
+     * reachable according to {@link #socketTester} and absent values for those not. Checks are concurrent
+     * and the elements in the Iterable are ordered according to their position in sockets.
+     */
+    private Iterable<Optional<HostAndPort>> tryReachable(Iterable<? extends HostAndPort> sockets, final Duration timeout) {
+        final List<ListenableFuture<Optional<HostAndPort>>> futures = Lists.newArrayList();
+        final AtomicReference<Stopwatch> sinceFirstCompleted = new AtomicReference<>();
+
         for (final HostAndPort socket : sockets) {
-            futures.add(userExecutor.submit(new Runnable() {
+            futures.add(userExecutor.submit(new Callable<Optional<HostAndPort>>() {
+                @Override
+                public Optional<HostAndPort> call() {
+                    // Whether the socket was reachable (vs. the result of call, which is whether the repeater is done).
+                    final AtomicBoolean theResultWeCareAbout = new AtomicBoolean();
+                    Repeater.create("socket-reachable")
+                            .limitTimeTo(timeout)
+                            .backoffTo(Duration.FIVE_SECONDS)
+                            .until(new Callable<Boolean>() {
+                                @Override
+                                public Boolean call() throws TimeoutException {
+                                    boolean reachable = socketTester.apply(socket);
```

Comment:
Socket connect timeout is OS dependent and could be well over a minute. This will trip the grace period timeout below.